src/fsharp/FSharp.Core/Linq.fs

```diff
@@ -355,141 +359,108 @@ module LeafExpressionConverter =
                 let props = ctor.DeclaringType.GetProperties()
                 Expression.New(ctor, argsR, [| for p in props -> (p :> MemberInfo) |]) |> asExpr
 
-
             // Do the same thing as C# compiler for string addition
-            | PlusQ (_, [ty1; ty2; ty3], [x1; x2]) when (ty1 = typeof<string>) && (ty2 = typeof<string>) && (ty3 = typeof<string>) ->
+            | PlusQ (_, m, [x1; x2]) when
+                 m.GetGenericArguments() = [|typeof<string>; typeof<string>; typeof<string>|] ->
```

Comment:
Please don't use equality with an array. The array will be allocated each time the equality check is performed.  Either write code for this like before or use an active pattern.

Reply:
The active patterns themselves convert arrays returned from reflection to lists, which  are quite some allocations. What about them?